les/tokensale.go

```diff
@@ -0,0 +1,870 @@
+	// from and the callback that is going to process the results.
+	lespayCmd struct {
+		cmd    []byte
+		id     enode.ID
+		freeID string
+		send   func([]byte, uint)
+	}
+)
+
+// priority returns the processing priority for the next command coming from the given
+// source. Commands sent before the previously recommended delay has elapsed have a
+// lower priority. It also checks whether the number of commands consecutively sent
+// before the delay has elapsed exceeds maxBurst and rejects the command instantly if
+// necessary.
+func (c *cmdSource) priority() (int64, bool) {
+	dt := c.delayUntil - mclock.Now()
+	if dt <= 0 {
+		c.burstCounter = 0
+		return 0, true
+	}
+	if c.burstCounter >= tsMaxBurst {
+		return 0, false
+	}
+	c.burstCounter++
+	return -int64(dt), true
+}
+
+// addDelay adds the given amount to the recommended delay
+func (c *cmdSource) addDelay(now mclock.AbsTime, delay time.Duration) uint {
+	dt := time.Duration(c.delayUntil - now)
+	if dt <= 0 {
+		dt = 0
+	}
+	dt += delay
+	if dt < tsMinDelay {
+		dt = tsMinDelay
+	}
+	c.delayUntil = now + mclock.AbsTime(dt)
+	return uint((dt + time.Second - 1) / time.Second)
+}
+
+// delayFactor calculates the amount added to the recommended delay after processing
+// a single command
+func (t *tokenSale) delayFactor(now mclock.AbsTime) time.Duration {
+	if now > t.delayFactorZero {
+		t.delayFactorZero = now
+	}
+	t.delayFactorZero += mclock.AbsTime(t.tsTargetPeriod) + t.delayFactorLast - now
+	t.delayFactorLast = now
+	if now >= t.delayFactorZero {
+		return 0
+	} else {
+		return time.Duration(t.delayFactorZero-now) / 4
+	}
+}
+
+// sourceMapCleanup removes unnecessary entries from the command source map
+func (t *tokenSale) sourceMapCleanup() {
+	t.qlock.Lock()
+	defer t.qlock.Unlock()
+
+	now := mclock.Now()
+	for src, s := range t.sources {
+		if s.delayUntil < now {
+			delete(t.sources, src)
+		}
+	}
+}
+
+// queueCommand schedules a lespay command (encapsulated in a lespayCmd) for execution
+func (t *tokenSale) queueCommand(src string, cmd lespayCmd) bool {
+	t.qlock.Lock()
+	defer t.qlock.Unlock()
+
+	s := t.sources[src]
+	if s == nil {
+		s = &cmdSource{}
+		t.sources[src] = s
+	}
+	if s.ch != nil {
+		select {
+		case s.ch <- cmd:
+			return true
+		default:
+			return false
+		}
+	}
+	s.ch = make(chan lespayCmd, 16)
+	s.ch <- cmd
+
+	go func() {
+	loop:
+		for {
+			select {
+			case cmd := <-s.ch:
+				t.qlock.Lock()
+				pri, ok := s.priority()
+				t.qlock.Unlock()
+				if ok {
+					task := t.sq.newTask(nil, 0, pri)
+					if !task.start() {
+						break loop
+					}
+					reply := t.runCommand(cmd.cmd, cmd.id, cmd.freeID)
+					t.qlock.Lock()
+					now := mclock.Now()
+					delay := s.addDelay(now, t.delayFactor(now))
+					t.qlock.Unlock()
+					cmd.send(reply, delay)
+					time.Sleep(t.tsProcessDelay)
+					task.done()
+				} else {
+					cmd.send(nil, 0)
+				}
+			default:
+				break loop
+			}
+			t.qlock.Lock()
+			s.ch = nil
+			t.qlock.Unlock()
```

Comment:
There is a data race:

- the loop tries to exit and wait the `qlock` to release channel.
- however is the lock is obtained by another `queueCommand`  call and the cmd is queued in the channel
- loop gets the lock and set nil
- new queued cmd is dropped silently